### ISSUE TITLE
spurious diagnostic: Variable `None` should have snake_case name

### ISSUE THREAD

**Issue by ConradIrwin** [rust-lang/rust-analyzer#19709]
**rust-analyzer version**: 2025-04-28 (and I think the few releases prior)

**rustc version**: rustc 1.86.0 (05f9846f8 2025-03-31)

**editor or extension**: Zed

**relevant settings**: None

**repository link (if public, optional)**: https://github.com/zed-industries/zed

**code snippet to reproduce**:
```rust
// add your code here

```

I haven't been able to figure out a minimal reproduction for this yet, but across the Zed team, we've noticed that sometimes (maybe 1-10% of the time?) when first starting Zed with rust-analyzer we'll get a spurious diagnostic:

![Image](https://github.com/user-attachments/assets/857799b8-7ea1-4e34-8daf-61b31e87a311)

It's also not clear how to get rid of this. Sometimes it seems to go away, other times I get frustrated and resort to `cargo clean` followed by restarting Zed.

It looks like it's not unique to Zed: https://github.com/rust-lang/rust-analyzer/issues/17759#issuecomment-2829939900.

I can keep looking out for things that seem to cause this, and am happy to try and test any hypothesis you have for why it might happen

**Comment by lnicola** [comment]
Do you have `rust-src` installed?

**Comment by Veykril** [comment]
cargo clean won't do anything. That diagnostic comes from r-a itself.

This should only happen if we compute diagnostics without the sysroot being loaded, as we then no longer resolve `None` to a constant, but to a binding instead. I think this can happen on startup when we build the crategraph while the sysroot sources haven't been loaded into the VFS yet (due to I/O).

It should disappear if you retrigger native diagnostics which (technically depends on the editor but) usually just editing the file in question should suffice.

**Comment by Veykril** [comment]
To be a bit more clear on the matter, the code here is the culprit https://github.com/rust-lang/rust-analyzer/blob/5adee2ad2cde49ad5a7d43f483c2884f48900c3b/crates/rust-analyzer/src/reload.rs#L725-L731

That returns an `Option<FileId>` and it is used to assign a file id to a crate's crate root. If that happens to return `None`, the crate gets discarded (for reasons) which causes the issues here if it is `core` or `std`, until a new VFS event comes in for one of those crate root files (`crate_graph_file_dependencies`) that kicks off a crate graph rebuilt.

**Comment by ConradIrwin** [comment]
@Veykril ok, that makes sense. I was about to file another issue where sometimes crates that change on disk (e.g. through `git checkout`) are sometimes not reloaded. Seems like this could actually be exactly the same issue though?

Would you be open to helping me fix this? Spurious diagnostics are a huge pain point for me (and the zed team as a whole) right now; but I don't know how the current system is supposed to work.

 I think we'll both be in Utretch in a week or two, otherwise I'm pretty open next week (on UK time).

**Comment by Veykril** [comment]
Changing the way the VFS works is definitely on our list and now that we have migrated to the new salsa it should be workable in fact. Happy to sit down at RustWeek and ponder about it a bit. I'd love to get this fixed as well.

> ok, that makes sense. I was about to file another issue where sometimes crates that change on disk (e.g. through git checkout) are sometimes not reloaded. Seems like this could actually be exactly the same issue though?

If it fixes itself after a bit without restarting r-a then yes very likely. If not, then that sounds like a file watching issue (which could be client or server depending on which is being used)

**Comment by ConradIrwin** [comment]
Great, thank you! Looking forward to chatting. 

Sounds like thestale crates is actually file-system watching, let me see what we're doing there.

**Comment by xandkar** [comment]
I noticed the same behavior in my `nvim` + `coc.nvim` + `rust-analyzer` setup. Started after one of the recent updates, not sure which one yet.

**Comment by ChayimFriedman2** [comment]
@xandkar What is your Rust version?

**Comment by xandkar** [comment]
> [@xandkar](https://github.com/xandkar) What is your Rust version?

```
$ rustc --version
rustc 1.85.0 (4d91de4e4 2025-02-17)
```

**Comment by xandkar** [comment]
> > [@xandkar](https://github.com/xandkar) What is your Rust version?
> 
> ```
> $ rustc --version
> rustc 1.85.0 (4d91de4e4 2025-02-17)
> ```

Update: doesn't seem to be happening with `1.86.0` 🤞

**Comment by ChayimFriedman2** [comment]
So we upgraded our Rust version again by mistake? @Veykril @lnicola do you know if we did something/something was done in rust-lang/rust that could lead to this?

**Comment by Veykril** [comment]
Not that I am aware of

**Comment by thisisscotts** [comment]
Chiming in to say that I experienced this issue in both VSCode and Zed on Windows. Minimally reproduced with:

```
fn main() {
    println!("Hello, world!");

    let nothing: Option<String> = None;

    match nothing {
        Some(_) => println!("Got something"),
        None => println!("Found nothing"),
    }
}
```

where "None" would be highlighted with a warning about snake case naming.

Other issues encountered included:

- Hovering over variables only to see that they were evaluated as type `{unknown}`
- No help pop-ups for methods
- No auto-suggestions/completions for available methods

I tried a number of suggestions elsewhere about reinstalling `rust-src`, changing my version of `rustc`, setting `RUST_SRC_PATH` and forcing `STABLE` toolchains to no avail.

Hints from [this thread](https://github.com/rust-lang/rust-analyzer/issues/18560) and [this other thread](https://github.com/rust-lang/rust-analyzer/issues/17759). Seems the main cause on my end is because of network issues, specifically something to do with corporate network restrictions imposed upon me. 

Fixed by changing my `C:\Users\<username here>\.cargo\config.toml` to include:

```
[http]
check-revoke = false
```
then restarting RA.

Note that setting the check-revoke to false is not recommended, as far as I have read, so I only use this when I need to update/download crates. I typically comment out the check revoke afterwards.

```
[http]
#check-revoke = false
```
Now that I have updated whatever cache or data store needed updating, I cannot reproduce the error. Highlighting, variable types and auto-suggestion etc are all working as normal now.

**Comment by franklaranja** [comment]
I have this all the time. The warning goes away when I fully qualify None. I am running the latest version if rust-analyzer provided by termux on termux and I work offline.

**Comment by Levi-Montgomery** [comment]
Had the same problem, I simply restarted my rust-analyzer extension and that fixed it.

**Comment by franklaranja** [comment]
Didn't work for me I put
use core::option::Option::None; // for rust-analyzer
in every file that has a None :(

**Comment by Levi-Montgomery** [comment]
Yeah i spoke to soon myself. It just came back the next day.

**Comment by qjerome** [comment]
I've been experiencing the exact same issue. **All the components were installed yet I still had the warning** and couldn't go to the `Option<T>` enum in rust-src code by hitting go to definition shortcut, **even though the rust-src files were on my disk**. 

As @franklaranja, I am working offline (isolated in a Linux namespace with no network).

**What solved my issue was**: temporarily allow networking and restart rust-analyzer.
Now even offline it works 👍. 

I can conclude from my experience that networking seems to be needed for rust-analyzer to properly analyze rust-src. What was weird was that I could indeed navigate to some of the structures in rust-src (String, Arc, Rc ...) but not Option<T>. So maybe the code analysis fails in the middle of something when networking isn't available.

If that is indeed a networking issue (would be cool if others could confirm), maybe it would be nice to address this in such a way that rust-src analysis is not failing if no network is available.

**Comment by ChayimFriedman2** [comment]
@qjerome Indeed, rust-analyzer needs to download dependencies of std.

**Comment by franklaranja** [comment]
Why does rust-analyzer need to get std stuff from the net whilst cargo/rustc don't? It would not only be nice for people without internet connection, but for everybody as it would  make rust-analyzer faster, if it could avoid unnecessary internet access.

**Comment by lnicola** [comment]
The compiler doesn't have to look at the source code of the standard library or let you use "Go to definition" on its symbols.

**Comment by franklaranja** [comment]
Are you sure? How do you cross compile? I have the source of std locally (I use termux which is a non standard way of running Rust.) Maybe rust analyzer can check first locally.

As a side note this can't be the real problem regarding None as it is the only item from std::prelude that causes problems.

**Comment by ShoyuVanilla** [comment]
> Are you sure? How do you cross compile? I have the source of std locally (I use termux which is a non standard way of running Rust.) Maybe rust analyzer can check first locally.
> 
> As a side note this can't be the real problem regarding None as it is the only item from std::prelude that causes problems.

You are misunderstanding. What needs to be downloaded are the "dependencies of standard library", not standard library itself. It is prebuilt and bundled with rustc but rust-analyzer cannot directly parse those artifacts

**Comment by flodiebold** [comment]
> As a side note this can't be the real problem regarding None as it is the only item from std::prelude that causes problems.

It's just one of the more prominent ways that a broken std shows in RA. You're guaranteed to also have other problems, e.g. missing completions.

**Comment by Levi-Montgomery** [comment]
can someone please share how to download "the dependencies of std" then

**Comment by ShoyuVanilla** [comment]
> can someone please share how to download "the dependencies of std" then

Actually, it is not handled by rust-analyzer directly. Rust-analyzer invokes cargo with the following commands and cargo fetches `crates.io` and those dependencies:

1. `cargo metadata`: This is a necessary step to correctly model the project. If this fails, rust-analyzer fallbacks to `cargo metadata --no-deps`, which does not require internet connection. In such case, the types defined in `std` itself like `Mutex` are resolved well, but types like `Option` are not, as they are re-exported items from other "dependencies" - in `Option`'s case, `core`.  
2. `cargo check --compile-time-deps`: I guess failing on this won't cause this issue.

So, we can't really help on manually fetching those dependencies instead of cargo and then doing some hacks to convince cargo that it has done deps fetch successfully.

The right approach would be diagnosing why `cargo metadata` fails in your machine and fixing the cause but that needs what the logs are saying.

I'm suspecting https://github.com/rust-lang/rust-analyzer/issues/20250 and planning to fix it soon with giving more notifications on `cargo metadata` failures.

**Comment by franklaranja** [comment]
Maybe pass the --offline to cargo, when offline?

**Comment by Levi-Montgomery** [comment]
I'm not even offline when getting the issue, and `cargo metadata` runs fine. Restarting extensions fixes it for a while but it always comes back.

Edit: upon further reflection it appears to be because my Mac has no internet when i close the lid. And rust-analyzer will do its little refresh thing while my laptop is asleep and not have internet. And then it doesn't make an attempt to try again when i open my laptop so i have to manually restart the extension while i have internet.

**Comment by traceflight** [comment]
I resolved this issue by updating some crates in my offline repository. These crates were mentioned in the error messages from  cargo metadata , which were displayed in the VS Code Output panel of the rust-analyzer language server.

**Comment by mibes404** [comment]
In my case this error was caused by having vendored sources. Once I removed the vendored sources from my .cargo/config.toml and restarted the extension, the issue disappeared. The issue can be consistently reproduced by running `cargo vendor`, after which the problem reappears.

**Comment by jacklund** [comment]
Just to add more cases, in my case it was that one of my dependent libraries had used `None` as the name of a variant for one of its enums (an enum which I didn't even use), but this seemed to have confused `rust_analyzer` enough to cause this issue for me.

Adding `core::option::Option::None` at the top where it was showing this fixed the issue, as mentioned earlier.

**Comment by stephanemagnenat** [comment]
I believe I have the same problem. I used the LLM to try to find a reproducible test case, which I believe it managed (see #23366). In my case it seems linked to Zed reporting deleted files as existing to rust-analyzer. See also zed-industries/zed#64231

Maybe my reproducer can be helpful to you: https://github.com/stephanemagnenat/ra-ghost-module

In my opinion this is a error reporting quality bug.
